src/level-editor/levelEditorPage.tsx

```diff
@@ -0,0 +1,147 @@
+import React, { ReactNode, useEffect, useRef } from 'react'
+import { useSelector, useDispatch } from '@dwalter/spider-hook'
+import {
+  tilesSelector,
+  addTile,
+  goatLocationSelector,
+  relocateGoat,
+  Location,
+  Tile as TileObject,
+  currentlyPlacingColorSelector,
+  cyclePlacingColor,
+  flipTile,
+} from './editorState'
+import { utils as SpiderUtils } from '@dwalter/spider-hook'
+
+const useEventListener = ((type: string, handler: Function, options: any) => {
+  const handlerRef = useRef(handler)
+
+  handlerRef.current = handler
+
+  const deps = [type, options]
+
+  const shouldUpdate = SpiderUtils.useShouldUpdate(deps)
+
+  useEffect(
+    shouldUpdate
+      ? () => {
+          const artificialHandler = (event: Event) => handlerRef.current(event)
+
+          addEventListener(type, artificialHandler, options)
+
+          return () => removeEventListener(type, artificialHandler, options)
+        }
+      : (null as any),
+    deps,
+  )
+}) as typeof addEventListener
+
+export function LevelEditorPage() {
+  const currentlyPlacingColor = useSelector(currentlyPlacingColorSelector)
+  const tiles = useSelector(tilesSelector)
+  const goatLocation = useSelector(goatLocationSelector)
+  const dispatch = useDispatch()
+
+  useEventListener('keydown', event => {
+    console.log(event.key)
+    switch (event.key) {
+      case 'w':
+        dispatch(relocateGoat([goatLocation[0], goatLocation[1] - 1]))
+        return
+
+      case 'a':
+        dispatch(relocateGoat([goatLocation[0] - 1, goatLocation[1]]))
+        return
+
+      case 's':
+        dispatch(relocateGoat([goatLocation[0], goatLocation[1] + 1]))
+        return
+
+      case 'd':
+        dispatch(relocateGoat([goatLocation[0] + 1, goatLocation[1]]))
+        return
+
+      case 'c':
```

Comment:
`c` cycles through colors, and `t` can do tile types. Not sure how to handle tiles that are the same color on the underside... (could just track a secondary color and secondary type with `v` and `y` or something)

Spacebar flips